controllers/cluster/cluster_controller.go

```diff
@@ -88,89 +88,22 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
 		}
 	}
 
-	log = log.WithValues("Cluster", cluster.Namespace+"/"+cluster.Name)
-
-	if _, exist := cluster.Labels[akoov1alpha1.AviClusterLabel]; !exist {
-		log.Info("Cluster doesn't have AVI enabled, skip Cluster reconciling")
-		return res, nil
-	}
-
-	log.Info("Cluster has AVI enabled, start Cluster reconciling")
-	// Getting all akodeploymentconfigs
-	var akoDeploymentConfigs akoov1alpha1.AKODeploymentConfigList
-	if err := r.Client.List(ctx, &akoDeploymentConfigs); err != nil {
+	akoDeploymentConfig, err := ako_operator.GetAKODeploymentConfigForCluster(ctx, r.Client, log, cluster)
+	if err != nil {
+		log.Error(err, "failed to get cluster matched akodeploymentconfig")
 		return res, err
 	}
 
-	// Matches current cluster with all the akoDeploymentConfigs
-	clusterLabels := cluster.GetLabels()
-	matchedAkoDeploymentConfigs := make([]akoov1alpha1.AKODeploymentConfig, 0)
-	for _, akoDeploymentConfig := range akoDeploymentConfigs.Items {
-		if selector, err := metav1.LabelSelectorAsSelector(&akoDeploymentConfig.Spec.ClusterSelector); err != nil {
-			log.Error(err, "Failed to convert label sector to selector when matching ", "cluster", cluster.Name, " with ", akoDeploymentConfig.Name)
-		} else if selector.Matches(labels.Set(clusterLabels)) {
-			log.Info("Cluster ", cluster.Name, " is selected by", "Akodeploymentconfig", (akoDeploymentConfig.Namespace + "/" + akoDeploymentConfig.Name))
-			matchedAkoDeploymentConfigs = append(matchedAkoDeploymentConfigs, akoDeploymentConfig)
-		}
-	}
-
-	// When the cluster is not selected or only selected by the default ADC, a.k.a. install-ako-for-all
-	// We drop its skip-default-adc label, allowing cluster to use default ADC's config.
-	// It happens when a cluster is no longer selected by a customized ADC
-	if len(matchedAkoDeploymentConfigs) == 0 ||
-		(len(matchedAkoDeploymentConfigs) == 1 && matchedAkoDeploymentConfigs[0].Name == akoov1alpha1.WorkloadClusterAkoDeploymentConfig) {
-		delete(cluster.Labels, akoov1alpha1.AviClusterSelectedLabel)
-	}
-
-	// If the cluster is selected by some AKODeploymentConfig objects, skip removing the finalizer
-	if len(matchedAkoDeploymentConfigs) > 0 {
+	if akoDeploymentConfig == nil {
+		// Removing finalizer if current cluster can't be selected by any akoDeploymentConfig
+		log.Info("Not find cluster matched akodeploymentconfig, removing finalizer", "finalizer", akoov1alpha1.ClusterFinalizer)
+		ctrlutil.RemoveFinalizer(cluster, akoov1alpha1.ClusterFinalizer)
+		// Removing avi label after deleting all the resources
+		delete(cluster.Labels, akoov1alpha1.AviClusterLabel)
+		log.Info("Cluster doesn't have AVI enabled, skip Cluster reconciling")
```

Comment:
No need to delete addon-secret?

Reply:
I realize that one is duplicated with what tanzu-addon-manager already does here: https://github.com/vmware-tanzu/tanzu-framework/blob/425d3b8a0132ec7cd1143c30469b2ff8c2a09ecb/addons/controllers/addon_controller.go#L176. 

When the addon secret got created, we kinda delegate to addon manager to manage the lifecycle of it for us.